aws/rust-runtime/aws-http/src/content_encoding.rs

```diff
@@ -0,0 +1,325 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+use bytes::Bytes;
+use http::{HeaderMap, HeaderValue};
+use http_body::{Body, SizeHint};
+use pin_project_lite::pin_project;
+
+use std::pin::Pin;
+use std::task::{Context, Poll};
+
+const CRLF: &str = "\r\n";
+const CHUNK_TERMINATOR: &str = "0\r\n";
+
+/// Content encoding header value constants
+pub mod header_value {
+    /// Header value denoting "aws-chunked" encoding
+    pub const AWS_CHUNKED: &str = "aws-chunked";
+}
+
+/// Options used when constructing an [`AwsChunkedBody`][AwsChunkedBody].
+#[derive(Debug, Default)]
+#[non_exhaustive]
+pub struct AwsChunkedBodyOptions {
+    /// The total size of the stream. Because we only support unsigned encoding
+    /// this implies that there will only be a single chunk containing the
+    /// underlying payload.
+    pub stream_length: u64,
+    /// The length of each trailer sent within an `AwsChunkedBody`. Necessary in
+    /// order to correctly calculate the total size of the body accurately.
+    pub trailer_lens: Vec<u64>,
+}
+
+impl AwsChunkedBodyOptions {
+    /// Create a new [`AwsChunkedBodyOptions`][AwsChunkedBodyOptions]
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Set stream length
+    pub fn with_stream_length(mut self, stream_length: u64) -> Self {
+        self.stream_length = stream_length;
+        self
+    }
+
+    /// Set a trailer len
+    pub fn with_trailer_len(mut self, trailer_len: u64) -> Self {
+        self.trailer_lens.push(trailer_len);
+        self
+    }
+}
+
+#[derive(Debug, PartialEq, Eq)]
+enum AwsChunkedBodyState {
+    /// Write out the size of the chunk that will follow. Then, transition into the
+    /// `WritingChunk` state.
+    WritingChunkSize,
+    /// Write out the next chunk of data. Multiple polls of the inner body may need to occur before
+    /// all data is written out. Once there is no more data to write, transition into the
+    /// `WritingTrailers` state.
+    WritingChunk,
+    /// Write out all trailers associated with this `AwsChunkedBody` and then transition into the
+    /// `Closed` state.
+    WritingTrailers,
+    /// This is the final state. Write out the body terminator and then remain in this state.
+    Closed,
+}
+
+pin_project! {
+    /// A request body compatible with `Content-Encoding: aws-chunked`
+    ///
+    /// Chunked-Body grammar is defined in [ABNF] as:
+    ///
+    /// ```txt
+    /// Chunked-Body    = *chunk
+    ///                   last-chunk
+    ///                   chunked-trailer
+    ///                   CRLF
+    ///
+    /// chunk           = chunk-size CRLF chunk-data CRLF
+    /// chunk-size      = 1*HEXDIG
+    /// last-chunk      = 1*("0") CRLF
+    /// chunked-trailer = *( entity-header CRLF )
+    /// entity-header   = field-name ":" OWS field-value OWS
```

Comment:
should explain somewhere that this will always write the request into a single chunk